src/dmd/root/string.d

```diff
@@ -212,3 +212,82 @@ unittest
     assert(dstrcmp("안녕하세요!", "안녕하세요!") == 0);
     static assert(dstrcmp("안녕하세요!", "안녕하세요!") == 0);
 }
+
+/**
+ * Infers the length `N` of a string literal and coerces its type to a static
+ * array with length `N + 1`. Returns the string with a null character appended
+ * to the end.
+ *
+ * Params:
+ *  literal = string literal
+ *
+ * Notes:
+ *  - LDC produces quite optimal code for short strings:
+ *    - https://d.godbolt.org/z/M69Z1g
+ *    - https://gist.github.com/PetarKirov/338e4ab9292b6b2b311a3070572a07fb (backup URL)
+*/
+T[N + 1] toStaticArray(T, size_t N)(scope const(T)[N] literal)
+if (is(T == char) || is(T == wchar) || is(T == dchar))
+{
+    T[N+1] result = void;
+    result[0..N] = literal[0..N];
+    result[N] = 0;
+    return result;
+}
+
+///
+@safe pure nothrow @nogc
+unittest
+{
+    auto m = "123".toStaticArray;
+    const c = "123".toStaticArray;
+    immutable i = "123".toStaticArray;
+    enum e = "123".toStaticArray;
+
+    assert(m == "123\0");
+    assert(c == "123\0");
+    assert(i == "123\0");
+    static assert(e == "123\0");
+
+    const empty = "".toStaticArray;
+    static assert(empty.length == 1);
+    static assert(empty[0] == '\0');
+}
+
+/**
+ * Checks if C string `p` starts with `needle`.
+ * Params:
+ *     p = the C string to check
+ *     needle = the string to look for
+ * Returns:
+ *    `true` if `p` starts with `needle`
+ */
+@system pure nothrow @nogc
+bool startsWith(scope const(char)* p, scope const(char)[] needle)
+in { assert(p && needle.ptr); }
+do
+{
+    foreach (const c; needle)
+    {
+        if (!*p || c != *p)
```

Comment:
Actually, the first `!*p` is not necessary. If `*p` is `0`, the second case will trip as long as `c` is not `0`, and it should never be `0`.

Reply:
On first look, I thought the same. However, I added it because it's possible (however unlikely) that the `needle` slice contains an embedded null character, which would cause `startsWith` to read out-of-bounds data:

``` D
static immutable str = "123";
const ptr = &str[0];
assert(!ptr.startsWith("123\0"));  // failed before
assert(!ptr.startsWith("123\04")); // out-of-bounds access before
```

Reply:
Except I can't see a circumstance with an embedded 0 in dmd. An assert for it would be better.

Reply:
> Except I can't see a circumstance with an embedded 0 in dmd.

In dmd no, but user input can't be trusted in general.

Reply:
User input from where?

Reply:
That's why I suggested replacing it with an assert.